File: dart_copy/lib/src/features/monitoring/domain/machine_state.dart
/// Machine and process state reported by the ESP32.
///
/// The specification (§3, §9) asks the app to show *what the machine is doing
/// right now*. That is two separate facts: the connection/run state of the
/// machine as a whole ([MachineStatus]) and the position of the batch in the
/// filtration sequence ([FiltrationStage]). Both arrive in every packet.
library;

/// Run state of the filtration machine — specification §3.
enum MachineStatus {
  /// The app has no link to the machine.
  disconnected('Disconnected', 'No link to the machine.'),

  /// Linked but not yet armed for a run.
  connected('Connected', 'Linked to the machine.'),

  /// Armed and waiting for the operator to start a batch.
  ready('Ready', 'Ready to start a batch.'),

  /// A filtration/assessment cycle is under way.
  running('Running', 'Filtration in progress.'),

  /// Cycle suspended; it can be resumed.
  paused('Paused', 'Cycle paused.'),

  /// Cycle finished normally.
  completed('Completed', 'Cycle complete.'),

  /// The machine reported a fault.
  error('Error', 'The machine reported a fault.');

  const MachineStatus(this.label, this.description);

  final String label;
  final String description;

  bool get isActive => this == running || this == paused;

  /// Parses the wire value (`"RUNNING"`, `"running"`, `"run"`, …).
  ///
  /// Unknown values fall back to [fallback] rather than throwing: firmware
  /// should be able to add states without crashing the app.
  static MachineStatus parse(
    Object? raw, {
    MachineStatus fallback = MachineStatus.connected,
  }) {
    if (raw == null) return fallback;
    final key = raw.toString().trim().toLowerCase().replaceAll(RegExp('[ _-]'), '');
    for (final status in MachineStatus.values) {
      if (status.name.toLowerCase() == key) return status;
    }
    return switch (key) {
      'idle' || 'standby' || 'waiting' => MachineStatus.ready,
      'run' || 'active' || 'busy' => MachineStatus.running,
      'done' || 'finished' || 'complete' => MachineStatus.completed,
      'halted' || 'stopped' || 'suspended' => MachineStatus.paused,
      'fault' || 'err' || 'failure' => MachineStatus.error,
      _ => fallback,
    };
  }
}

/// Position of the batch in the filtration sequence — specification §9.
enum FiltrationStage {
  idle('Idle', 'Machine idle'),
  extracting('Extracting', 'Drawing honey into the system'),
  primaryFiltration('Primary filtration', 'Coarse filtration'),
  secondaryFiltration('Secondary filtration', 'Fine filtration'),
  qualityAssessment('Quality assessment', 'Measuring quality parameters'),
  finalTransfer('Final transfer', 'Transferring filtered honey'),
  completed('Completed', 'Batch complete'),
  paused('Paused', 'Sequence paused'),
  error('Error', 'Sequence halted by a fault');

  const FiltrationStage(this.label, this.description);

  final String label;
  final String description;

  /// The ordered run of stages a normal batch passes through. [paused] and
  /// [error] are excluded — they interrupt the sequence rather than advance it.
  static const List<FiltrationStage> sequence = <FiltrationStage>[
    idle,
    extracting,
    primaryFiltration,
    secondaryFiltration,
    qualityAssessment,
    finalTransfer,
    completed,
  ];

  bool get isTerminal => this == completed || this == error;

  /// How far through [sequence] this stage sits, 0.0–1.0. Stages outside the
  /// sequence report the progress of the run they interrupted as unknown.
  double? get progress {
    final index = sequence.indexOf(this);
    if (index < 0) return null;
    return index / (sequence.length - 1);
  }

  static FiltrationStage parse(
    Object? raw, {
    FiltrationStage fallback = FiltrationStage.idle,
  }) {
    if (raw == null) return fallback;
    final key = raw.toString().trim().toLowerCase().replaceAll(RegExp('[ _-]'), '');
    for (final stage in FiltrationStage.values) {
      if (stage.name.toLowerCase() == key) return stage;
    }
    return switch (key) {
      'extract' || 'extraction' || 'intake' => FiltrationStage.extracting,
      'primary' || 'filtration1' || 'coarsefiltration' =>
        FiltrationStage.primaryFiltration,
      'secondary' || 'filtration2' || 'finefiltration' =>
        FiltrationStage.secondaryFiltration,
      'assessment' || 'quality' || 'testing' =>
        FiltrationStage.qualityAssessment,
      'transfer' || 'final' || 'dispensing' => FiltrationStage.finalTransfer,
      'done' || 'finished' || 'complete' => FiltrationStage.completed,
      _ => fallback,
    };
  }
}

File: react_native/src/features/monitoring/domain/machineState.ts
/**
 * Machine and process state reported by the ESP32.
 *
 * The specification (§3, §9) asks the app to show *what the machine is doing
 * right now*. That is two separate facts: the connection/run state of the
 * machine as a whole ([MachineStatus]) and the position of the batch in the
 * filtration sequence ([FiltrationStage]). Both arrive in every packet.
 */

/** Run state of the filtration machine — specification §3. */
export type MachineStatus =
  | 'disconnected'
  | 'connected'
  | 'ready'
  | 'running'
  | 'paused'
  | 'completed'
  | 'error';

export const MACHINE_STATUSES: readonly MachineStatus[] = [
  'disconnected',
  'connected',
  'ready',
  'running',
  'paused',
  'completed',
  'error',
];

const MACHINE_STATUS_INFO: Record<MachineStatus, { label: string; description: string }> = {
  disconnected: { label: 'Disconnected', description: 'No link to the machine.' },
  connected: { label: 'Connected', description: 'Linked to the machine.' },
  ready: { label: 'Ready', description: 'Ready to start a batch.' },
  running: { label: 'Running', description: 'Filtration in progress.' },
  paused: { label: 'Paused', description: 'Cycle paused.' },
  completed: { label: 'Completed', description: 'Cycle complete.' },
  error: { label: 'Error', description: 'The machine reported a fault.' },
};

export const machineStatusLabel = (s: MachineStatus) => MACHINE_STATUS_INFO[s].label;
export const machineStatusDescription = (s: MachineStatus) => MACHINE_STATUS_INFO[s].description;
export const isMachineActive = (s: MachineStatus) => s === 'running' || s === 'paused';

const normaliseKey = (raw: unknown) => String(raw).trim().toLowerCase().replace(/[ _-]/g, '');

/**
 * Parses the wire value (`"RUNNING"`, `"running"`, `"run"`, …).
 *
 * Unknown values fall back to [fallback] rather than throwing: firmware
 * should be able to add states without crashing the app.
 */
export function parseMachineStatus(
  raw: unknown,
  fallback: MachineStatus = 'connected',
): MachineStatus {
  if (raw == null) return fallback;
  const key = normaliseKey(raw);
  for (const status of MACHINE_STATUSES) {
    if (status.toLowerCase() === key) return status;
  }
  switch (key) {
    case 'idle':
    case 'standby':
    case 'waiting':
      return 'ready';
    case 'run':
    case 'active':
    case 'busy':
      return 'running';
    case 'done':
    case 'finished':
    case 'complete':
      return 'completed';
    case 'halted':
    case 'stopped':
    case 'suspended':
      return 'paused';
    case 'fault':
    case 'err':
    case 'failure':
      return 'error';
    default:
      return fallback;
  }
}

/** Position of the batch in the filtration sequence — specification §9. */
export type FiltrationStage =
  | 'idle'
  | 'extracting'
  | 'primaryFiltration'
  | 'secondaryFiltration'
  | 'qualityAssessment'
  | 'finalTransfer'
  | 'completed'
  | 'paused'
  | 'error';

export const FILTRATION_STAGES: readonly FiltrationStage[] = [
  'idle',
  'extracting',
  'primaryFiltration',
  'secondaryFiltration',
  'qualityAssessment',
  'finalTransfer',
  'completed',
  'paused',
  'error',
];

const STAGE_INFO: Record<FiltrationStage, { label: string; description: string }> = {
  idle: { label: 'Idle', description: 'Machine idle' },
  extracting: { label: 'Extracting', description: 'Drawing honey into the system' },
  primaryFiltration: { label: 'Primary filtration', description: 'Coarse filtration' },
  secondaryFiltration: { label: 'Secondary filtration', description: 'Fine filtration' },
  qualityAssessment: { label: 'Quality assessment', description: 'Measuring quality parameters' },
  finalTransfer: { label: 'Final transfer', description: 'Transferring filtered honey' },
  completed: { label: 'Completed', description: 'Batch complete' },
  paused: { label: 'Paused', description: 'Sequence paused' },
  error: { label: 'Error', description: 'Sequence halted by a fault' },
};

export const stageLabel = (s: FiltrationStage) => STAGE_INFO[s].label;
export const stageDescription = (s: FiltrationStage) => STAGE_INFO[s].description;

/**
 * The ordered run of stages a normal batch passes through. `paused` and
 * `error` are excluded — they interrupt the sequence rather than advance it.
 */
export const STAGE_SEQUENCE: readonly FiltrationStage[] = [
  'idle',
  'extracting',
  'primaryFiltration',
  'secondaryFiltration',
  'qualityAssessment',
  'finalTransfer',
  'completed',
];

export const isStageTerminal = (s: FiltrationStage) => s === 'completed' || s === 'error';

/**
 * How far through the sequence this stage sits, 0.0–1.0. Stages outside the
 * sequence report the progress of the run they interrupted as unknown.
 */
export function stageProgress(stage: FiltrationStage): number | null {
  const index = STAGE_SEQUENCE.indexOf(stage);
  if (index < 0) return null;
  return index / (STAGE_SEQUENCE.length - 1);
}

export function parseFiltrationStage(
  raw: unknown,
  fallback: FiltrationStage = 'idle',
): FiltrationStage {
  if (raw == null) return fallback;
  const key = normaliseKey(raw);
  for (const stage of FILTRATION_STAGES) {
    if (stage.toLowerCase() === key) return stage;
  }
  switch (key) {
    case 'extract':
    case 'extraction':
    case 'intake':
      return 'extracting';
    case 'primary':
    case 'filtration1':
    case 'coarsefiltration':
      return 'primaryFiltration';
    case 'secondary':
    case 'filtration2':
    case 'finefiltration':
      return 'secondaryFiltration';
    case 'assessment':
    case 'quality':
    case 'testing':
      return 'qualityAssessment';
    case 'transfer':
    case 'final':
    case 'dispensing':
      return 'finalTransfer';
    case 'done':
    case 'finished':
    case 'complete':
      return 'completed';
    default:
      return fallback;
  }
}
